File: internal/nano/tensor.go
// Package nano is a from-scratch, dependency-free transformer language model in pure Go:
// a real autograd engine + GPT + trainer. It is "the smallest model we can actually build
// and verify in-session" — trained from zero on CPU, no GPU, no cgo, no downloads at runtime.
//
// Design alignment: nano implements the kernel.Kernel seam (see internal/kernel/nano.go), so the
// same tiny-llm system that will later host a Qwen3 Cortex via ggml can host this native Cortex today.
package nano

import (
	"math"
	"runtime"
	"sync"
)

// Tensor is a 2D [R,C] float32 array with a gradient and a backward closure.
type Tensor struct {
	Data []float32
	Grad []float32
	R, C int
	bw   func() // accumulates this node's Grad into its parents' Grads
}

// Tape records intermediate tensors in creation (topological) order so Backward can
// replay their backward closures in reverse. Parameters are leaves and live off-tape.
type Tape struct{ nodes []*Tensor }

// NewTape returns an empty tape for one forward/backward pass.
func NewTape() *Tape { return &Tape{} }

func (tp *Tape) node(r, c int) *Tensor {
	t := &Tensor{Data: make([]float32, r*c), Grad: make([]float32, r*c), R: r, C: c}
	tp.nodes = append(tp.nodes, t)
	return t
}

// Backward seeds d(loss)/d(loss)=1 and replays every node's backward in reverse order.
func (tp *Tape) Backward(loss *Tensor) {
	loss.Grad[0] = 1
	for i := len(tp.nodes) - 1; i >= 0; i-- {
		if tp.nodes[i].bw != nil {
			tp.nodes[i].bw()
		}
	}
}

// NewParam creates a persistent leaf parameter tensor (kept across steps).
func NewParam(r, c int) *Tensor {
	return &Tensor{Data: make([]float32, r*c), Grad: make([]float32, r*c), R: r, C: c}
}

// Leaf wraps existing data as a non-tracked input (e.g. token indices, masks).
func Leaf(data []float32, r, c int) *Tensor {
	return &Tensor{Data: data, Grad: make([]float32, r*c), R: r, C: c}
}

// ZeroGrad clears a parameter's gradient buffer between steps.
func (t *Tensor) ZeroGrad() {
	for i := range t.Grad {
		t.Grad[i] = 0
	}
}

// nWorkers bounds matmul parallelism; leave 2 cores for the OS/runtime (research 00/05).
var nWorkers = max(1, runtime.NumCPU()-2)

// parFor runs fn over [0,n) split across up to nWorkers goroutines.
func parFor(n int, fn func(i int)) {
	if n == 0 {
		return
	}
	w := nWorkers
	if w > n {
		w = n
	}
	if w == 1 {
		for i := 0; i < n; i++ {
			fn(i)
		}
		return
	}
	var wg sync.WaitGroup
	chunk := (n + w - 1) / w
	for g := 0; g < w; g++ {
		lo := g * chunk
		hi := lo + chunk
		if hi > n {
			hi = n
		}
		if lo >= hi {
			break
		}
		wg.Add(1)
		go func(lo, hi int) {
			defer wg.Done()
			for i := lo; i < hi; i++ {
				fn(i)
			}
		}(lo, hi)
	}
	wg.Wait()
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func geluTanh(x float32) float32 {
	// 0.5x(1+tanh(√(2/π)(x+0.044715x³)))
	x3 := x * x * x
	inner := 0.7978845608 * (x + 0.044715*x3)
	return 0.5 * x * (1 + float32(math.Tanh(float64(inner))))
}

func geluGrad(x float32) float32 {
	x3 := x * x * x
	inner := 0.7978845608 * (x + 0.044715*x3)
	t := float32(math.Tanh(float64(inner)))
	sech2 := 1 - t*t
	dinner := 0.7978845608 * (1 + 3*0.044715*x*x)
	return 0.5*(1+t) + 0.5*x*sech2*dinner
}

File: src/nano/tensor.rs
//! The autograd core: a 2D float32 tensor with a gradient, and a Tape that records each
//! op's backward closure so `backward` can replay them in reverse.
//!
//! The Go original used a pointer graph (`*Tensor` with a `bw func()` closure per node). This
//! Rust port models the same graph with `Rc<RefCell<Tensor>>` handles captured by boxed backward
//! closures stored on the Tape, recorded in creation (topological) order.
//!
//! Note on parallelism: the Go engine parallelized matmuls across goroutines (`parFor`). Each
//! output element's reduction is computed by a single worker, so results are independent of worker
//! count. This port computes those loops sequentially — numerically identical and fully
//! deterministic — trading some throughput for a safe, allocation-free translation.

use std::cell::RefCell;
use std::rc::Rc;

/// Tensor is a 2D [r,c] float32 array with a gradient buffer.
pub struct Tensor {
    pub data: Vec<f32>,
    pub grad: Vec<f32>,
    pub r: usize,
    pub c: usize,
}

/// A shared, interior-mutable handle to a tensor (the node type of the autograd graph).
pub type T = Rc<RefCell<Tensor>>;

impl Tensor {
    /// ZeroGrad clears this tensor's gradient buffer between steps.
    pub fn zero_grad(&mut self) {
        for g in self.grad.iter_mut() {
            *g = 0.0;
        }
    }
}

/// Tape records each op's backward closure in creation order so [`Tape::backward`] can replay
/// them in reverse. Parameters are leaves and live off-tape.
pub struct Tape {
    pub(crate) bws: Vec<Box<dyn Fn()>>,
}

impl Default for Tape {
    fn default() -> Self {
        Tape::new()
    }
}

impl Tape {
    /// NewTape returns an empty tape for one forward/backward pass.
    pub fn new() -> Tape {
        Tape { bws: Vec::new() }
    }

    /// node allocates a fresh tracked tensor (zeroed data + grad).
    pub(crate) fn node(r: usize, c: usize) -> T {
        Rc::new(RefCell::new(Tensor {
            data: vec![0.0; r * c],
            grad: vec![0.0; r * c],
            r,
            c,
        }))
    }

    pub(crate) fn push_bw(&mut self, f: Box<dyn Fn()>) {
        self.bws.push(f);
    }

    /// Backward seeds d(loss)/d(loss)=1 and replays every recorded backward in reverse order.
    pub fn backward(&self, loss: &T) {
        loss.borrow_mut().grad[0] = 1.0;
        for bw in self.bws.iter().rev() {
            bw();
        }
    }
}

/// NewParam creates a persistent leaf parameter tensor (kept across steps).
pub fn new_param(r: usize, c: usize) -> T {
    Rc::new(RefCell::new(Tensor {
        data: vec![0.0; r * c],
        grad: vec![0.0; r * c],
        r,
        c,
    }))
}

/// Leaf wraps existing data as a non-tracked input (e.g. token indices, masks).
#[allow(dead_code)]
pub fn leaf(data: Vec<f32>, r: usize, c: usize) -> T {
    let n = r * c;
    Rc::new(RefCell::new(Tensor {
        data,
        grad: vec![0.0; n],
        r,
        c,
    }))
}

/// GELU activation (tanh approximation).
pub(crate) fn gelu_tanh(x: f32) -> f32 {
    // 0.5x(1+tanh(√(2/π)(x+0.044715x³)))
    let x3 = x * x * x;
    let inner = 0.7978845608_f32 * (x + 0.044715_f32 * x3);
    0.5 * x * (1.0 + (inner as f64).tanh() as f32)
}

pub(crate) fn gelu_grad(x: f32) -> f32 {
    let x3 = x * x * x;
    let inner = 0.7978845608_f32 * (x + 0.044715_f32 * x3);
    let t = (inner as f64).tanh() as f32;
    let sech2 = 1.0 - t * t;
    let dinner = 0.7978845608_f32 * (1.0 + 3.0 * 0.044715_f32 * x * x);
    0.5 * (1.0 + t) + 0.5 * x * sech2 * dinner
}
